### PR TITLE
Adds handling for specialized generics.

### PR DESCRIPTION
**Pull request by peterschutt**
### Pull Request Checklist

- [ ] New code has 100% test coverage
- [ ] (If applicable) The prose documentation has been updated to reflect the changes introduced by this PR
- [ ] (If applicable) The reference documentation has been updated to reflect the changes introduced by this PR
- [ ] Pre-Commit Checks were ran and passed
- [ ] Tests were ran and passed

### Description

-

### Close Issue(s)

-

### CODE REVIEW

**Comment by peterschutt** [comment]
Hey, I did some experimenting with some edge cases and thought might just be easier to PR this stuff in.

File: litestar/contrib/pydantic/pydantic_dto_factory.py [hunk 1]
```diff
@@ -37,6 +37,8 @@
     OpenAPIType.STRING: str,
     OpenAPIType.INTEGER: int,
     OpenAPIType.BOOLEAN: bool,
+    OpenAPIType.OBJECT: str,
+    OpenAPIType.NUMBER: float,
```
Comment (peterschutt) [inline litestar/contrib/pydantic/pydantic_dto_factory.py:41]: These are the openapi types that are in the `PYDANTIC_TYPE_MAP` that we didn't have down types for.

File: litestar/contrib/pydantic/pydantic_dto_factory.py [hunk 2]
```diff
@@ -72,8 +74,18 @@ def generate_field_definitions(
...
+            # field_definition.annotation will be `dict[str, Any]` and the metadata will be
+            # `(Json(),)` - that is, the metadata will be an _instance_ of `Json`, not the
+            # type itself. So we check for that case and use the type of the metadata instead
+            # of the annotation, if it exists in the `PYDANTIC_TYPE_MAP`.
+            annotation = next(
+                (type(t) for t in field_definition.metadata if type(t) in PYDANTIC_TYPE_MAP),
+                field_definition.annotation,
+            )
+            if annotation in PYDANTIC_TYPE_MAP:
+                open_api_type = PYDANTIC_TYPE_MAP[annotation].type
```
Comment (peterschutt) [inline litestar/contrib/pydantic/pydantic_dto_factory.py:88]: In pydantic v2, if you have a model like

```py
class MyModel(BaseModel):
    data: Json[dict[str, Any]]
```
The _actual_ annotation that gets added to the model is
```py
class MyModel(BaseModel):
    data: Annotated[dict[str, Any], Json()]
```
You can see these shenanigans here: https://github.com/pydantic/pydantic/blob/4dd40aec7397418e37b1029a0496b486adecbb7c/pydantic/types.py#L1375-L1377*

**Review by jw-can — APPROVED** [review]
thanks!
